File: src/main/scala/jp/ne/opt/redshiftfake/postgres/FakePostgresqlDriver.scala
package jp.ne.opt.redshiftfake.postgres

import java.sql.{DriverManager, Connection}
import java.util.Properties

import jp.ne.opt.redshiftfake.{Global, FakeConnection}
import jp.ne.opt.redshiftfake.s3.S3ServiceImpl
import org.postgresql.Driver

class FakePostgresqlDriver extends Driver {
  import FakePostgresqlDriver._

  override def connect(url: String, info: Properties): Connection = {
    new FakeConnection(
      DriverManager.getConnection(url.replaceFirst(urlPrefix, "jdbc:postgresql"), info),
      new S3ServiceImpl(Global.s3Endpoint))
  }

  override def acceptsURL(url: String): Boolean = url.startsWith(urlPrefix)
}

object FakePostgresqlDriver {
  private[this] val driverInstance = new FakePostgresqlDriver
  DriverManager.registerDriver(driverInstance)

  private val urlPrefix = "jdbc:postgresqlredshift"

  def createConnection(url: String, prop: java.util.Properties): Connection = {
    DriverManager.getConnection(url, prop)
  }
}

File: src/main/java/jp/ne/opt/redshiftfake/postgres/FakePostgresqlDriver.java
package jp.ne.opt.redshiftfake.postgres;

import jp.ne.opt.redshiftfake.FakeConnection;
import jp.ne.opt.redshiftfake.Global;
import jp.ne.opt.redshiftfake.s3.S3ServiceImpl;
import jp.ne.opt.redshiftfake.views.SystemViews;
import org.postgresql.Driver;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;

public class FakePostgresqlDriver extends Driver {
    private static final String urlPrefix = "jdbc:postgresqlredshift";

    static {
        try {
            DriverManager.registerDriver(new FakePostgresqlDriver());
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public Connection connect(String url, Properties info) throws SQLException {
        if (url.startsWith(urlPrefix)) {

            final String postgresUrl = url.replaceFirst(urlPrefix, "jdbc:postgresql");
            final Connection connection = DriverManager.getConnection(postgresUrl, info);
            SystemViews.create(connection);

            return new FakeConnection(
                    connection,
                    new S3ServiceImpl(Global.s3Endpoint())
            );
        } else {
            return null;
        }
    }

    @Override
    public boolean acceptsURL(String url) {
        return url.startsWith(urlPrefix);
    }
}
